src/api/controllers/v2/module.controller.ts

```diff
@@ -0,0 +1,142 @@
+  try {
+    const modules = await ModuleDataLayer.getAll();
+    if (!modules) res.status(404).send();
+    else res.status(200).send(modules);
+  } catch (e) {
+    res.status(500).send(e);
+  }
+};
+
+export const getModuleByID = async (
+  req: Request,
+  res: Response
+): Promise<void> => {
+  // add validator for moduleId?
+  try {
+    const module_ = await ModuleDataLayer.getById(req.params.moduleId);
+    if (!module_) res.status(404).send();
+    else res.status(200).send(module_);
+  } catch (e) {
+    res.status(500).send(e);
+  }
+};
+
+export const getModuleByProblemId = async (
+  req: Request,
+  res: Response
+): Promise<Record<string, any>> => {
+  // validate problemId?
+  // if (!isMongoId(req.params.problemId)) {
+  //   return res.status(400).send('This route requires a valid problem ID');
+  // }
+  const problem = await ProblemDataLayer.getById(req.params.problemId);
+  if (!problem) return res.status(400).send('invalid problem id');
+  try {
+    const modules = await ModuleDataLayer.getById(problem.moduleId);
+    if (!modules) return res.status(404).send();
+    return res.status(200).send(modules);
+  } catch (e) {
+    return res.status(500).send(e);
+  }
+};
+
+export const putModule = async (
+  req: Request,
+  res: Response
+): Promise<Record<string, any>> => {
+  try {
+    const result = await ModuleDataLayer.updateById(
+      req.params.moduleId,
+      req.body
+    );
+    return res.status(200).send(result);
+  } catch (e) {
+    return res.status(500).send(e);
+  }
+};
+
+export const deleteModule = async (
+  req: Request,
+  res: Response
+): Promise<Record<string, any>> => {
+  try {
+    const result = await ModuleDataLayer.deleteById(req.params.moduleId);
+    await ProblemDataLayer.deleteByModule(req.params.moduleId);
+    await LessonDataLayer.deleteByModule(req.params.moduleId);
+    return res.status(200).send(result);
+  } catch (e) {
+    return res.status(500).send(e);
+  }
+};
+
+export const changeProblemOrder = async (
+  req: Request,
+  res: Response
+): Promise<Record<string, any>> => {
+  try {
+    const module_ = await ModuleDataLayer.getById(req.params.moduleId);
+    if (!module_ || !module_['problems'])
+      return res.status(400).send('Module not found or module is empty');
+    if (module_['problems'].length == 1)
+      return res
+        .status(400)
+        .send('Cannot change order of only problem in module');
+
+    const problemIndex = module_['problems']
+      .map((problem: ProblemAttributes) => {
+        return problem.id;
+      })
+      .indexOf(req.body.problemId);
+    if (problemIndex == -1)
+      return res.status(400).send('Problem not found in module');
+
+    if (req.body.direction === 'up') {
+      if (problemIndex === 0)
+        return res.status(400).send('Problem already at top of module');
+
+      const problemToSwap = module_['problems'][problemIndex - 1];
+      module_['problems'][problemIndex - 1] = req.body.problemId;
```

Comment:
are we swapping problem ids here or the order ids? we should change the order ids and not the problem id as this can be a problem if someone is using links as the ids will now point to different problems.

Reply:
my bad, I copied that from v1 without thinking it through. should be fixed

Reply:
I think this is something Blake should fix because he wrote that logic initially. I can ask him to do it in the Fall once he is back.